contracts/galactic-express/src/lib.rs

```diff
@@ -122,322 +131,292 @@ impl Contract {
-                            fuel_left,
-                            payload: participant.payload,
-                        });
-                    }
-                    Err(halt_reason) => {
-                        actor_turns.push(TurnOutcome::Destroyed(halt_reason));
+        match turn_index {
+            0 => {
+                if random.chance(3) {
+                    return Err(HaltReason::EngineFailure);
+                }
 
-                        break;
-                    }
+                if fuel_left >= 80 - 2 * self.weather as u8 && random.chance(10) {
+                    return Err(HaltReason::FuelOverload);
                 }
             }
+            1 => {
+                if payload_amount >= 80 - 2 * self.weather as u8 && random.chance(10) {
+                    return Err(HaltReason::PayloadOverload);
+                }
 
-            turns.insert(*actor, actor_turns);
+                if random.chance(5 + self.weather as u8) {
+                    return Err(HaltReason::SeparationFailure);
+                }
+            }
+            2 => {
+                if random.chance(10 + self.weather as u8) {
+                    return Err(HaltReason::AsteroidCollision);
+                }
+            }
+            _ => unreachable!(),
         }
 
-        let mut scores: Vec<_> = turns
-            .iter()
-            .map(|(actor, turns)| {
-                let last_turn = turns.last().expect("there must be at least 1 turn");
-
-                (
-                    *actor,
-                    match last_turn {
-                        TurnOutcome::Alive { fuel_left, payload } => {
-                            *payload as u32 * self.altitude as u32 + *fuel_left as u32
-                        }
-                        TurnOutcome::Destroyed(_) => 0,
-                    },
-                )
-            })
-            .collect();
+        Ok(new_fuel_left)
+    }
 
-        scores.sort_by(|(_, score_a), (_, score_b)| score_a.cmp(score_b));
+    async fn start_game(&mut self, participant: Participant) -> Result<Event, Error> {
+        self.assert_admin(msg::source())?;
 
-        let mut rankings = Vec::with_capacity(scores.len());
-        let mut deductible = 0;
+        if self.participants.is_empty() {
+            return Err(Error::NotEnoughParticipants);
+        }
 
-        for (recipient, _) in scores.into_iter().rev() {
-            let amount = self.reward - deductible;
+        self.register(participant, false)?;
 
-            if FTokenEvent::Ok
-                != msg::send_for_reply_as(
-                    self.sft,
-                    FTokenAction::Message {
-                        transaction_id: stepper.step()?,
-                        payload: LogicAction::Mint { recipient, amount },
-                    },
-                    0,
-                    0,
-                )?
-                .await?
+        let mut random = Random::new()?;
+        let mut turns = vec![];
```

Comment:
Use [`Vec::with_capacity`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity) if you know the size of the output vector (`TOTAL_TURNS` as I see).